File: 15_Interest-Rate Counterparty Exposure and XVA/python/xva_adjustments.py
"""
Simplified CVA, DVA, and FVA calculations.

The implementation uses deterministic discount factors,
flat hazard rates, constant recovery rates, and discrete
exposure integration.
"""

import numpy as np


class CreditCurve:
    """
    Flat-hazard credit curve.
    """

    def __init__(
        self,
        hazard_rate,
        recovery_rate
    ):
        self.hazard_rate = float(
            hazard_rate
        )

        self.recovery_rate = float(
            recovery_rate
        )

        self.validate()

    def validate(
        self
    ):
        """
        Validate credit-curve parameters.
        """

        if self.hazard_rate < 0.0:
            raise ValueError(
                "hazard_rate must be non-negative."
            )

        if not (
            0.0
            <=
            self.recovery_rate
            <=
            1.0
        ):
            raise ValueError(
                "recovery_rate must lie between zero and one."
            )

    def survival_probability(
        self,
        times
    ):
        """
        Calculate survival probability.
        """

        times = np.asarray(
            times,
            dtype=float
        )

        return np.exp(
            -self.hazard_rate
            *
            times
        )

    def default_probability_increments(
        self,
        times
    ):
        """
        Calculate marginal default probabilities between dates.
        """

        times = np.asarray(
            times,
            dtype=float
        )

        survival = self.survival_probability(
            times
        )

        previous_survival = np.concatenate(
            (
                np.array(
                    [
                        1.0
                    ]
                ),
                survival[:-1]
            )
        )

        default_increments = (
            previous_survival
            -
            survival
        )

        return np.maximum(
            default_increments,
            0.0
        )

    def loss_given_default(
        self
    ):
        """
        Return one minus recovery.
        """

        return (
            1.0
            -
            self.recovery_rate
        )


def interpolate_initial_discount_factors(
    model,
    times
):
    """
    Interpolate initial discount factors to exposure dates.
    """

    times = np.asarray(
        times,
        dtype=float
    )

    if hasattr(
        model,
        "initial_discount_factors"
    ):
        discount_factors = np.asarray(
            model.initial_discount_factors,
            dtype=float
        )

    elif hasattr(
        model,
        "initialDiscountFactors"
    ):
        discount_factors = np.asarray(
            model.initialDiscountFactors,
            dtype=float
        )

    else:
        raise AttributeError(
            "The model must contain initial discount factors."
        )

    if hasattr(
        model,
        "tenor"
    ):
        tenor_times = np.asarray(
            model.tenor.times,
            dtype=float
        )

    elif hasattr(
        model,
        "tenor_structure"
    ):
        tenor_times = np.asarray(
            model.tenor_structure.times,
            dtype=float
        )

    else:
        raise AttributeError(
            "The model must contain tenor information."
        )

    return np.interp(
        times,
        tenor_times,
        discount_factors
    )


def calculate_cva(
    times,
    expected_exposure,
    discount_factors,
    counterparty_credit_curve
):
    """
    Calculate unilateral CVA.

    CVA is returned as a positive valuation cost.
    """

    times = np.asarray(
        times,
        dtype=float
    )

    expected_exposure = np.asarray(
        expected_exposure,
        dtype=float
    )

    discount_factors = np.asarray(
        discount_factors,
        dtype=float
    )

    default_increments = (
        counterparty_credit_curve
        .default_probability_increments(
            times
        )
    )

    cva = (
        counterparty_credit_curve
        .loss_given_default()
        *
        np.sum(
            discount_factors
            *
            expected_exposure
            *
            default_increments
        )
    )

    return float(
        cva
    )


def calculate_dva(
    times,
    expected_negative_exposure,
    discount_factors,
    bank_credit_curve
):
    """
    Calculate DVA as a positive valuation benefit.
    """

    times = np.asarray(
        times,
        dtype=float
    )

    expected_negative_exposure = np.asarray(
        expected_negative_exposure,
        dtype=float
    )

    discount_factors = np.asarray(
        discount_factors,
        dtype=float
    )

    default_increments = (
        bank_credit_curve
        .default_probability_increments(
            times
        )
    )

    dva = (
        bank_credit_curve
        .loss_given_default()
        *
        np.sum(
            discount_factors
            *
            expected_negative_exposure
            *
            default_increments
        )
    )

    return float(
        dva
    )


def calculate_funding_adjustments(
    times,
    expected_exposure,
    expected_negative_exposure,
    discount_factors,
    borrowing_spread,
    lending_spread=0.0
):
    """
    Calculate simple funding cost and benefit adjustments.

    FCA represents the cost of funding positive exposure.

    FBA represents a benefit associated with negative exposure.

    FVA is reported as:

        FVA = FCA - FBA
    """

    times = np.asarray(
        times,
        dtype=float
    )

    ee = np.asarray(
        expected_exposure,
        dtype=float
    )

    ene = np.asarray(
        expected_negative_exposure,
        dtype=float
    )

    discount_factors = np.asarray(
        discount_factors,
        dtype=float
    )

    borrowing_spread = float(
        borrowing_spread
    )

    lending_spread = float(
        lending_spread
    )

    if borrowing_spread < 0.0:
        raise ValueError(
            "borrowing_spread must be non-negative."
        )

    if lending_spread < 0.0:
        raise ValueError(
            "lending_spread must be non-negative."
        )

    if len(
        times
    ) == 1:
        time_steps = np.array(
            [
                0.0
            ]
        )

    else:
        time_steps = np.diff(
            times,
            prepend=times[0]
        )

    fca = np.sum(
        discount_factors
        *
        ee
        *
        borrowing_spread
        *
        time_steps
    )

    fba = np.sum(
        discount_factors
        *
        ene
        *
        lending_spread
        *
        time_steps
    )

    return {
        "FCA": float(
            fca
        ),
        "FBA": float(
            fba
        ),
        "FVA": float(
            fca
            -
            fba
        )
    }


def calculate_xva(
    model,
    exposure_metrics,
    counterparty_credit_curve,
    bank_credit_curve,
    borrowing_spread,
    lending_spread=0.0
):
    """
    Calculate CVA, DVA, FCA, FBA, FVA, and net adjustment.
    """

    times = exposure_metrics.times

    discount_factors = (
        interpolate_initial_discount_factors(
            model=model,
            times=times
        )
    )

    cva = calculate_cva(
        times=times,
        expected_exposure=(
            exposure_metrics.expected_exposure
        ),
        discount_factors=discount_factors,
        counterparty_credit_curve=(
            counterparty_credit_curve
        )
    )

    dva = calculate_dva(
        times=times,
        expected_negative_exposure=(
            exposure_metrics
            .expected_negative_exposure
        ),
        discount_factors=discount_factors,
        bank_credit_curve=bank_credit_curve
    )

    funding = calculate_funding_adjustments(
        times=times,
        expected_exposure=(
            exposure_metrics.expected_exposure
        ),
        expected_negative_exposure=(
            exposure_metrics
            .expected_negative_exposure
        ),
        discount_factors=discount_factors,
        borrowing_spread=borrowing_spread,
        lending_spread=lending_spread
    )

    net_adjustment = (
        -cva
        +
        dva
        -
        funding[
            "FVA"
        ]
    )

    return {
        "CVA": cva,
        "DVA": dva,
        "FCA": funding["FCA"],
        "FBA": funding["FBA"],
        "FVA": funding["FVA"],
        "net_valuation_adjustment": float(
            net_adjustment
        )
    }

File: 15_Interest-Rate Counterparty Exposure and XVA/cpp/XVAAdjustments.cpp
#include "XVAAdjustments.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>
#include <stdexcept>


CreditCurve::CreditCurve(
    double hazardRate,
    double recoveryRate
)
    :
    hazardRate_(hazardRate),
    recoveryRate_(recoveryRate)
{
    validate();
}


double CreditCurve::hazardRate() const
{
    return hazardRate_;
}


double CreditCurve::recoveryRate() const
{
    return recoveryRate_;
}


double CreditCurve::lossGivenDefault() const
{
    return
        1.0
        -
        recoveryRate_;
}


double CreditCurve::survivalProbability(
    double time
) const
{
    return std::exp(
        -hazardRate_
        *
        time
    );
}


Vector CreditCurve::survivalProbabilities(
    const Vector& times
) const
{
    Vector result(
        times.size(),
        0.0
    );

    for (std::size_t index = 0;
         index < times.size();
         ++index)
    {
        result[index] =
            survivalProbability(
                times[index]
            );
    }

    return result;
}


Vector CreditCurve::defaultProbabilityIncrements(
    const Vector& times
) const
{
    Vector survival =
        survivalProbabilities(
            times
        );

    Vector increments(
        times.size(),
        0.0
    );

    double previousSurvival =
        1.0;

    for (std::size_t index = 0;
         index < times.size();
         ++index)
    {
        increments[index] =
            std::max(
                previousSurvival
                -
                survival[index],
                0.0
            );

        previousSurvival =
            survival[index];
    }

    return increments;
}


void CreditCurve::validate() const
{
    if (hazardRate_ < 0.0)
    {
        throw std::invalid_argument(
            "hazardRate must be non-negative."
        );
    }

    if (recoveryRate_ < 0.0 ||
        recoveryRate_ > 1.0)
    {
        throw std::invalid_argument(
            "recoveryRate must lie between zero and one."
        );
    }
}


void XVAResult::printSummary() const
{
    std::cout
        << "\n"
        << "XVA Summary\n"
        << "------------------------------\n"
        << "CVA: "
        << cva
        << "\n"
        << "DVA: "
        << dva
        << "\n"
        << "FCA: "
        << fca
        << "\n"
        << "FBA: "
        << fba
        << "\n"
        << "FVA: "
        << fva
        << "\n"
        << "Net valuation adjustment: "
        << netValuationAdjustment
        << "\n";
}


double calculateCVA(
    const Vector& times,
    const Vector& expectedExposure,
    const Vector& discountFactors,
    const CreditCurve& counterpartyCreditCurve
)
{
    if (times.size() !=
            expectedExposure.size()
        ||
        times.size() !=
            discountFactors.size())
    {
        throw std::invalid_argument(
            "CVA input dimensions are inconsistent."
        );
    }

    Vector defaultIncrements =
        counterpartyCreditCurve
        .defaultProbabilityIncrements(
            times
        );

    double cva =
        0.0;

    for (std::size_t index = 0;
         index < times.size();
         ++index)
    {
        cva +=
            discountFactors[index]
            *
            expectedExposure[index]
            *
            defaultIncrements[index];
    }

    return
        counterpartyCreditCurve
        .lossGivenDefault()
        *
        cva;
}


double calculateDVA(
    const Vector& times,
    const Vector& expectedNegativeExposure,
    const Vector& discountFactors,
    const CreditCurve& bankCreditCurve
)
{
    if (times.size() !=
            expectedNegativeExposure.size()
        ||
        times.size() !=
            discountFactors.size())
    {
        throw std::invalid_argument(
            "DVA input dimensions are inconsistent."
        );
    }

    Vector defaultIncrements =
        bankCreditCurve
        .defaultProbabilityIncrements(
            times
        );

    double dva =
        0.0;

    for (std::size_t index = 0;
         index < times.size();
         ++index)
    {
        dva +=
            discountFactors[index]
            *
            expectedNegativeExposure[index]
            *
            defaultIncrements[index];
    }

    return
        bankCreditCurve
        .lossGivenDefault()
        *
        dva;
}


void calculateFundingAdjustments(
    const Vector& times,
    const Vector& expectedExposure,
    const Vector& expectedNegativeExposure,
    const Vector& discountFactors,
    double borrowingSpread,
    double lendingSpread,
    double& fca,
    double& fba,
    double& fva
)
{
    if (borrowingSpread < 0.0 ||
        lendingSpread < 0.0)
    {
        throw std::invalid_argument(
            "Funding spreads must be non-negative."
        );
    }

    if (times.size() !=
            expectedExposure.size()
        ||
        times.size() !=
            expectedNegativeExposure.size()
        ||
        times.size() !=
            discountFactors.size())
    {
        throw std::invalid_argument(
            "Funding input dimensions are inconsistent."
        );
    }

    fca = 0.0;
    fba = 0.0;

    for (std::size_t index = 1;
         index < times.size();
         ++index)
    {
        double dt =
            times[index]
            -
            times[index - 1];

        fca +=
            discountFactors[index]
            *
            expectedExposure[index]
            *
            borrowingSpread
            *
            dt;

        fba +=
            discountFactors[index]
            *
            expectedNegativeExposure[index]
            *
            lendingSpread
            *
            dt;
    }

    fva =
        fca
        -
        fba;
}


Vector interpolateDiscountFactors(
    const Vector& tenorTimes,
    const Vector& initialDiscountFactors,
    const Vector& exposureTimes
)
{
    if (tenorTimes.size() !=
        initialDiscountFactors.size())
    {
        throw std::invalid_argument(
            "tenorTimes and discount factors "
            "must have equal size."
        );
    }

    if (tenorTimes.empty())
    {
        throw std::invalid_argument(
            "tenorTimes cannot be empty."
        );
    }

    Vector result(
        exposureTimes.size(),
        0.0
    );

    for (std::size_t exposureIndex = 0;
         exposureIndex <
             exposureTimes.size();
         ++exposureIndex)
    {
        double target =
            exposureTimes[exposureIndex];

        if (target <=
            tenorTimes.front())
        {
            result[exposureIndex] =
                initialDiscountFactors.front();

            continue;
        }

        if (target >=
            tenorTimes.back())
        {
            result[exposureIndex] =
                initialDiscountFactors.back();

            continue;
        }

        for (std::size_t tenorIndex = 0;
             tenorIndex + 1 <
                 tenorTimes.size();
             ++tenorIndex)
        {
            double leftTime =
                tenorTimes[tenorIndex];

            double rightTime =
                tenorTimes[tenorIndex + 1];

            if (target >= leftTime &&
                target <= rightTime)
            {
                double weight =
                    (
                        target
                        -
                        leftTime
                    )
                    /
                    (
                        rightTime
                        -
                        leftTime
                    );

                result[exposureIndex] =
                    initialDiscountFactors[tenorIndex]
                    *
                    (1.0 - weight)
                    +
                    initialDiscountFactors[tenorIndex + 1]
                    *
                    weight;

                break;
            }
        }
    }

    return result;
}


XVAResult calculateXVA(
    const Vector& tenorTimes,
    const Vector& initialDiscountFactors,
    const ExposureMetricsResult& exposureMetrics,
    const CreditCurve& counterpartyCreditCurve,
    const CreditCurve& bankCreditCurve,
    double borrowingSpread,
    double lendingSpread
)
{
    Vector discountFactors =
        interpolateDiscountFactors(
            tenorTimes,
            initialDiscountFactors,
            exposureMetrics.times
        );

    XVAResult result;

    result.cva =
        calculateCVA(
            exposureMetrics.times,
            exposureMetrics.expectedExposure,
            discountFactors,
            counterpartyCreditCurve
        );

    result.dva =
        calculateDVA(
            exposureMetrics.times,
            exposureMetrics.expectedNegativeExposure,
            discountFactors,
            bankCreditCurve
        );

    calculateFundingAdjustments(
        exposureMetrics.times,
        exposureMetrics.expectedExposure,
        exposureMetrics.expectedNegativeExposure,
        discountFactors,
        borrowingSpread,
        lendingSpread,
        result.fca,
        result.fba,
        result.fva
    );

    result.netValuationAdjustment =
        -result.cva
        +
        result.dva
        -
        result.fva;

    return result;
}
